### PR TITLE
[llvm-dwp] Fix merging of debug_str_offsets with multiple contributions

### PR DESCRIPTION
**Pull request by molar**
This pull request will change the merging of ``debug_str_offset`` to merge per contribution and correctly copy over each contribution header to the merged section. I have added some test data which is in dwarf5 format as this is where the section contribution header was introduced, as far as i can tell.

### CODE REVIEW

**Comment by molar** [comment]
Ping @dwblaikie  kindly requesting a review

**Comment by dwblaikie** [comment]
Thanks!

Generally I think we're moving away from checked in binary object files - could you make the test from checked in assembly instead (perhaps using @MaskRay's new https://llvm.org/docs/TestingGuide.html#elaborated-tests perhaps)

So if I'm understanding correctly - you're saying llvm-dwp did not work when an input was a DWP and was DWARFv5? huh, surprising oversight... oh, I guess not that many people use dwp files as input to dwp actions, so I can see how some people would've been working along fine with this state of affairs.

**Comment by felipepiovezan** [comment]
I _think_ the dwarf YAML support should work for debug_str_offsets too?
 
> Generally I think we're moving away from checked in binary object files - could you make the test from checked in assembly instead (perhaps using @MaskRay's new https://llvm.org/docs/TestingGuide.html#elaborated-tests perhaps)

File: llvm/test/tools/llvm-dwp/X86/merge_v5.test [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+CHECK: 0x[[#%.8x,COFF:]]:
+CHECK-LABEL: Compile Unit: length = {{.*}}, format = DWARF32, version = 0x0005, unit_type = DW_UT_split_compile, abbr_offset =
+CHECK:         0x[[CAOFF]], addr_size = 0x08, DWO_id = 0x[[DWOC:.*]] (next unit at 0x[[#%.8x,BOFF:]])
+CHECK: [[#BOFF]]:
+CHECK-LABEL: Compile Unit: length = {{.*}}, format = DWARF32, version = 0x0005, unit_type = DW_UT_split_compile, abbr_offset =
+CHECK:         0x[[BAOFF]], addr_size = 0x08, DWO_id = 0x[[DWOB:.*]] (next unit at 0x[[#%.8x,AOFF:]])
+CHECK: [[#AOFF]]:
+CHECK-LABEL: Compile Unit: length = {{.*}}, format = DWARF32, version = 0x0005, unit_type = DW_UT_split_compile, abbr_offset =
+CHECK:         0x[[AAOFF]], addr_size = 0x08, DWO_id = 0x[[DWOA:.*]] (next unit at 0x[[#%.8x,XOFF:]])
+
```
Comment (dwblaikie) [inline llvm/test/tools/llvm-dwp/X86/merge_v5.test:27]: Given that the DWP tool doesn't modify the abbrev and info sections at all, probably omit these from the check? The dump of just the str_offsets section is, I think, enoguh to demonstrate the issue/fix, yeah?

File: llvm/test/tools/llvm-dwp/X86/merge_v5.test [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+CHECK:    Index Signature          INFO                                     ABBREV                                  STR_OFFSETS 
+CHECK-DAG:       [[DWOC]]  [0x00000000[[#COFF]], 0x00000000[[#BOFF]]) [0x0000[[CAOFF]], 0x0000[[AAOFF]]) [0x00000000, 0x00000024) 
+CHECK-DAG:       [[DWOB]]  [0x00000000[[#BOFF]], 0x00000000[[#AOFF]]) [0x0000[[BAOFF]],  
+CHECK-DAG:       [[DWOA]]  [0x00000000[[#AOFF]], 0x00000000[[#XOFF]]) [0x0000[[AAOFF]], 
+
+CHECK-LABEL: .debug_str_offsets.dwo contents:
+CHECK: Contribution size = 32, Format = DWARF32, Version = 5
+CHECK: Contribution size = 24, Format = DWARF32, Version = 5
+CHECK: Contribution size = 28, Format = DWARF32, Version = 5
+
```
Comment (dwblaikie) [inline llvm/test/tools/llvm-dwp/X86/merge_v5.test:38]: So the old behavior would've been a single contribution, and now there's three? /might/ be worth checking the content too, but I wouldn't insist on it/I could see it going either way, the length probably provides a fair amount of checking (like we didn't emit all the strings into one contribution, and then emit two empty headers)

Comment (molar) [reply llvm/test/tools/llvm-dwp/X86/merge_v5.test]: I have added a verbose check of the contents, let me know if there is a better way to do it, other than just comparing against the output.

**Comment by molar** [comment]
> Thanks!
> 
> Generally I think we're moving away from checked in binary object files - could you make the test from checked in assembly instead (perhaps using @MaskRay's new https://llvm.org/docs/TestingGuide.html#elaborated-tests perhaps)
> 
> So if I'm understanding correctly - you're saying llvm-dwp did not work when an input was a DWP and was DWARFv5? huh, surprising oversight... oh, I guess not that many people use dwp files as input to dwp actions, so I can see how some people would've been working along fine with this state of affairs.

Thank you for looking at this PR. 

We encounter this bug when using bazel along with llvm-dwp tool to generate .dwp files for our top-level binaries. In bazel it will switch to a tree merging when the number of .dwo inputs is larger than 100. It does not use the `-e` flag together with the binary which does the correct thing.  

https://github.com/bazelbuild/bazel/blob/7048ef5eb5395679241f2321a8104b26ecee01c0/src/main/starlark/builtins_bzl/common/cc/cc_debug_helper.bzl#L129

**Review by dwblaikie — APPROVED** [review]
Looks good, thanks!

**Comment by MaskRay** [comment]
(Might be useful to use `[DWP] Fix ...` or `[llvm-dwp] Fix ...` for the title)

**Comment by dwblaikie** [comment]
Do you need someone to merge this for you?

**Comment by molar** [comment]
> Do you need someone to merge this for you?

Yes please

**Pull request merged**